drivers/syslog/syslog_channel.c

```diff
@@ -234,23 +229,31 @@ g_syslog_channel[CONFIG_SYSLOG_MAX_CHANNELS] =
 #ifdef CONFIG_SYSLOG_DEFAULT
 static int syslog_default_putc(FAR syslog_channel_t *channel, int ch)
 {
-  UNUSED(channel);
-
 #  ifdef CONFIG_ARCH_LOWPUTC
+  /* See https://github.com/apache/nuttx/issues/14662
+   * about what this critical section is for.
+   */
+
+  irqstate_t flags = enter_critical_section();
```

Comment:
```suggestion
  irqstate_t flags = spin_lock_irqsave(&g_lowputs_lock);
```
could we replace critical section to spinlock_irqsave to get better performance?

Reply:
let's merge this patch since it fix the default syslog self-output issue and unify serial and syslog in next.

Reply:
> Even if it is replaced with critical section, the problem is not fundamentally resolved. The syslog printout will be mixed with printf, because serial.c is a character driver, and the backend needs to ensure that all characters can be output only when `uart_txready()` is reached. I think we need to unify the `/dev/console` output behavior with syslog, and keep a unified protection mechanism (lock/irq) in single-line printing.

my motivation of this PR is a block-forever-unexpectedly issue. (https://github.com/apache/nuttx/issues/14662)
this is not about messages-are-intermixed issues.

having said that, for a longer term, i tend to agree we should use the ordinary drivers/serial.c stuff (or whatever interrupt-based i/o) for syslog mainly because polling-based i/o like up_uputc would be simply slower.